### PR TITLE
HAL mul8x8to16 added

### LINKED ISSUE #25034: Extend OpenCV HAL multiply with signed and unsigned cases
### Describe the feature and motivation

examples : 8u->16u, 8s->16s. It allows to get rid of extra convertTo call to get signed values

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by savuor**
fixes #25034

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/src/hal_replacement.hpp [hunk 1]
```diff
@@ -298,6 +298,8 @@ inline int hal_ni_mul16s(const short *src1_data, size_t src1_step, const short *
 inline int hal_ni_mul32s(const int *src1_data, size_t src1_step, const int *src2_data, size_t src2_step, int *dst_data, size_t dst_step, int width, int height, double scale) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
 inline int hal_ni_mul32f(const float *src1_data, size_t src1_step, const float *src2_data, size_t src2_step, float *dst_data, size_t dst_step, int width, int height, double scale) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
 inline int hal_ni_mul64f(const double *src1_data, size_t src1_step, const double *src2_data, size_t src2_step, double *dst_data, size_t dst_step, int width, int height, double scale) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
+inline int hal_ni_mul8uExtend(const uchar* src1_data, size_t src1_step, const uchar* src2_data, size_t src2_step, ushort* dst_data, size_t dst_step, int width, int height, double scale) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
```
Comment (vpisarev) [inline modules/core/src/hal_replacement.hpp:301]: usually, IPP and other libs use notation like 8u16u to describe the situation when input and output types are different. Can we avoid `Extend` in the name and use `8u16u` suffix?

Comment (asmorkalov) [reply modules/core/src/hal_replacement.hpp:301]: I vote for 8u16u. Agreed with Vadim's concern, but it requires significant internals redesign and HAL API redesign.

Comment (asmorkalov) [reply modules/core/src/hal_replacement.hpp]: Applied the comment to PR.

File: modules/core/src/arithm.cpp [hunk 2]
```diff
@@ -947,7 +947,60 @@ void cv::copyTo(InputArray _src, OutputArray _dst, InputArray _mask)
...
+{
+    bool works = true;
+
+    // check that HAL functions are presented and work properly
+    uchar ua = 0, ub = 0;
+    ushort uc = 0;
+    int res;
+    res = cv_hal_mul8u16u(/* src1_data */ &ua, /* src1_step */ 1, /* src2_data */ &ub, /* src2_step */ 1,
+                             /* dst_data */ &uc, /* dst_step */ 1, /* width */ 1, /* height */ 1, /* scale */ 1);
+    if (res == CV_HAL_ERROR_NOT_IMPLEMENTED)
```
Comment (asmorkalov) [inline modules/core/src/arithm.cpp:978]: The HAL function may return not implemented status, if size is to small or data is not aligned. Static pre-check does not work.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/core/src/arithm.cpp [hunk 3]
```diff
@@ -947,7 +947,25 @@ void cv::copyTo(InputArray _src, OutputArray _dst, InputArray _mask)
...
+    cv_hal_mul8u16u(src1, step1, src2, step2, (ushort*)dst, step, width, height, scale);
+}
+
+static void mul8s16sWrapper(const uchar* src1, size_t step1,
+                               const uchar* src2, size_t step2,
+                               uchar* dst, size_t step, int width, int height,
+                               void* usrdata)
+{
+    double scale = *((double*)usrdata);
+    cv_hal_mul8s16s((schar*)src1, step1, (schar*)src2, step2, (short*)dst, step, width, height, scale);
```
Comment (asmorkalov) [inline modules/core/src/arithm.cpp:965]: `cv_hal_mul8s16s` and `cv_hal_mul8u16u` return value is not checked. They may return error or not implemented status.

File: modules/core/src/arithm.cpp [hunk 4]
```diff
@@ -984,12 +1007,19 @@ static BinaryFuncC* getRecipTab()
...
 
 void multiply(InputArray src1, InputArray src2,
-                  OutputArray dst, double scale, int dtype)
+              OutputArray dst, double scale, int dtype)
 {
     CV_INSTRUMENT_REGION();
 
-    arithm_op(src1, src2, dst, noArray(), dtype, getMulTab(),
-              true, &scale, std::abs(scale - 1.0) < DBL_EPSILON ? OCL_OP_MUL : OCL_OP_MUL_SCALE);
+    static bool halMul8to16available = (cv_hal_mul8u16u != hal_ni_mul8u16u) && (cv_hal_mul8s16s != hal_ni_mul8s16s);
```
Comment (asmorkalov) [inline modules/core/src/arithm.cpp:1014]: What if HAL implements only one of them?

Comment (savuor) [reply modules/core/src/arithm.cpp]: Fixed

File: modules/core/src/arithm.cpp (hunk 3 above)
Comment (savuor) [reply modules/core/src/arithm.cpp]: Have an ugly solution for this, please take a look

**Comment by vpisarev** [comment]
Can we replace `getMulTab(bool extended)` function with `getMulFunc(int src1depth, int src2depth, int dstdepth)`?
We can pass such a function into arithm_op. This function will first try to get direct function for the original src1, src2 and dst depths and if it fails, it will make another try with normalized types (in the latter case the explicit conversion `8u->32f` is done for each of the inputs in the case of `8u x 8u -> 32f` op).

File: modules/core/src/arithm.cpp [hunk 5]
```diff
@@ -947,6 +970,38 @@ void cv::copyTo(InputArray _src, OutputArray _dst, InputArray _mask)
...
 {
 
+static int mul8u16uWrapper(const uchar* src1, size_t step1,
+                           const uchar* src2, size_t step2,
+                           uchar* dst, size_t step, int width, int height,
+                           void* usrdata)
+{
+    double scale = *((double*)usrdata);
+    int res = cv_hal_mul8u16u(src1, step1, src2, step2, (ushort *)dst, step, width, height, scale);
+    if (res == 0 || res == CV_HAL_ERROR_NOT_IMPLEMENTED)
```
Comment (asmorkalov) [inline modules/core/src/arithm.cpp:980]: Use constant instead of 0.

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**